sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.orc
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.hadoop.io._
+import org.apache.orc.mapred.{OrcList, OrcMap, OrcStruct, OrcTimestamp}
+import org.apache.orc.storage.serde2.io.{DateWritable, HiveDecimalWritable}
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.SpecificInternalRow
+import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+private[orc] class OrcDeserializer(
+    dataSchema: StructType,
+    requiredSchema: StructType,
+    maybeMissingSchema: Option[StructType]) {
+
+  private[this] val mutableRow = new SpecificInternalRow(requiredSchema.map(_.dataType))
+
+  private[this] val valueWrappers = requiredSchema.fields.map(f => getValueWrapper(f.dataType))
+
+  def deserialize(writable: OrcStruct): InternalRow = {
+    convertOrcStructToInternalRow(writable, dataSchema, requiredSchema,
```

Comment:
can you follow the code style in `OrcFileFormat.unwrapOrcStructs`? Basically create an unwrapper for each field, and unwrapper is a `(Any, InternalRow, Int) => Unit`

Reply:
your implementation here doesn't consider boxing for primitive types at all.

Reply:
We use `valueWrapper` for each field [here](https://github.com/apache/spark/pull/19651/files/f644c6a88b4f24376c67028d0e927a2ee49fedbe#diff-b933819e6dcaff41eee8fce1e8f2932cR40). Do you mean changing name?

Reply:
Your wrapper returns a value, while the old implementation's wrapper set value to `InternalRow`, which avoids boxing.

Reply:
Oh. I see.